storm-core/src/jvm/org/apache/storm/zookeeper/LeaderElectorImp.java

```diff
@@ -0,0 +1,135 @@
+
+    public LeaderElectorImp(Map conf, List<String> servers, CuratorFramework zk, String leaderlockPath, String id, AtomicReference<LeaderLatch> leaderLatch,
+                            AtomicReference<LeaderLatchListener> leaderLatchListener) {
+        this.conf = conf;
+        this.servers = servers;
+        this.zk = zk;
+        this.leaderLatch = leaderLatch;
+        this.id = id;
+        this.leaderLatch = leaderLatch;
+        this.leaderLatchListener = leaderLatchListener;
+    }
+
+    @Override
+    public void prepare(Map conf) {
+        LOG.info("no-op for zookeeper implementation");
+    }
+
+    @Override
+    public void addToLeaderLockQueue() {
+        // if this latch is already closed, we need to create new instance.
+        try {
+            if (LeaderLatch.State.CLOSED.equals(leaderLatch.get().getState())) {
+                leaderLatch.set(new LeaderLatch(zk, leaderlockPath));
+                leaderLatchListener.set(Zookeeper.leaderLatchListenerImpl(conf, zk, leaderLatch.get()));
+            }
+            // Only if the latch is not already started we invoke start
+            if (LeaderLatch.State.LATENT.equals(leaderLatch.get().getState())) {
+                leaderLatch.get().addListener(leaderLatchListener.get());
+                leaderLatch.get().start();
+                LOG.info("Queued up for leader lock.");
+            } else {
+                LOG.info("Node already in queue for leader lock.");
+            }
+        } catch (Exception e) {
+            LOG.warn("failed to add LeaderLatchListener", e);
+        }
+
+    }
+
+    @Override
+    // Only started latches can be closed.
+    public void removeFromLeaderLockQueue() {
+        if (LeaderLatch.State.STARTED.equals(leaderLatch.get().getState())) {
+            try {
+                leaderLatch.get().close();
+                LOG.info("Removed from leader lock queue.");
+            } catch (IOException e) {
+                LOG.warn("failed to  close leaderLatch!!");
+            }
+
+        } else {
+            LOG.info("leader latch is not started so no removeFromLeaderLockQueue needed.");
+        }
+    }
+
+    @Override
+    public boolean isLeader() throws Exception {
+        return leaderLatch.get().hasLeadership();
+    }
+
+    @Override
+    public NimbusInfo getLeader() {
+        try {
+            return Zookeeper.toNimbusInfo(leaderLatch.get().getLeader());
+        } catch (Exception e) {
+            throw Utils.wrapInRuntime(e);
+        }
+    }
+
+    @Override
+    public List<NimbusInfo> getAllNimbuses() {
+        List<NimbusInfo> nimbusInfos = new ArrayList<>();
+        try {
+            Collection<Participant> participants = leaderLatch.get().getParticipants();
+            for (Participant participant : participants) {
+                nimbusInfos.add(Zookeeper.toNimbusInfo(participant));
+            }
+        } catch (Exception e) {
+            LOG.warn("failed to get nimbuses", e);
+        }
```

Comment:
Again don't eat the exception throw it, and wrap it if you cannot update the interface.